src/libraries/System.Text.Json/Common/ReflectionExtensions.cs

```diff
@@ -229,5 +230,85 @@ public static bool IsVirtual(this PropertyInfo? propertyInfo)
             Debug.Assert(propertyInfo != null);
             return propertyInfo != null && (propertyInfo.GetMethod?.IsVirtual == true || propertyInfo.SetMethod?.IsVirtual == true);
         }
+
+        public static bool IsKeyValuePair(this Type type, Type? keyValuePairType = null)
+        {
+            if (!type.IsGenericType)
+            {
+                return false;
+            }
+
+            // Work around not being able to use typeof(KeyValuePair<,>) directly during compile-time src gen type analysis.
+            keyValuePairType ??= typeof(KeyValuePair<,>);
+
+            Type generic = type.GetGenericTypeDefinition();
+            return generic == keyValuePairType;
+        }
+
+        public static bool TryGetDeserializationConstructor(
+#if !BUILDING_SOURCE_GENERATOR
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)]
+#endif
+            this Type type,
+            bool useDefaultCtorInAnnotatedStructs,
+            out ConstructorInfo? deserializationCtor)
+        {
+            ConstructorInfo? ctorWithAttribute = null;
+            ConstructorInfo? publicParameterlessCtor = null;
+            ConstructorInfo? lonePublicCtor = null;
+
+            ConstructorInfo[] constructors = type.GetConstructors(BindingFlags.Public | BindingFlags.Instance);
+
+            if (constructors.Length == 1)
```

Comment:
Seems like `lonePublicCtor` can be deferred until the `if` below

Reply:
I tried to refactor this method a bit (on top of this simple change to share the logic between src-gen and non src-gen), but kept hitting failing tests. I prefer to do this in an isolated PR if that's okay.